src/Lykke.SettingsReader/SettingsProcessor/SettingsProcessor.cs

```diff
@@ -217,43 +223,76 @@ private static string CheckProperty<T>(PropertyInfo property, T model)
                         break;
                 }
 
-                foreach (string val in valuesToCheck)
-                {
-                    if (string.IsNullOrWhiteSpace(val))
-                    {
-                        var optionalAttribute = property.GetCustomAttribute(typeof(OptionalAttribute));
-                        if (optionalAttribute == null)
-                            throw new CheckFieldException(property.Name, val, "Empty setting value");
-                        continue;
-                    }
+                tasks = valuesToCheck.Select(val => DoCheck(model, property, checker, val)).ToList();
 
-                    var checkResult = checker.CheckField(model, property.Name, val);
-                    Console.WriteLine(checkResult.Description);
-                    if (!checkResult.Result && checkResult.ThrowExceptionOnFail)
-                        return checkResult.Description;
-                }
+                errorMessages = (await Task.WhenAll(tasks))
+                    .Where(item => !string.IsNullOrEmpty(item))
+                    .ToList();
+
+                return errorMessages.Count == 0 ? null : string.Join(Environment.NewLine, errorMessages);
             }
-            else if (property.CanWrite)
+
+            if (!property.CanWrite) 
+                return null;
+            
+            object[] values = GetValuesToCheck(property, model);
+
+            tasks = values.Select(ProcessChecks).ToList();
+
+            errorMessages = (await Task.WhenAll(tasks))
+                .Where(item => !string.IsNullOrEmpty(item))
+                .ToList();
+            
+            return errorMessages.Count == 0 ? null : string.Join(Environment.NewLine, errorMessages);
+        }
+
+        private static async Task<string> DoCheck<T>(T model, PropertyInfo property, ISettingsFieldChecker checker, string val)
+        {
+            if (string.IsNullOrWhiteSpace(val))
             {
-                object[] values = GetValuesToCheck(property, model);
-                var errorMessages = new List<string>();
-                Parallel.ForEach(values, val =>
-                {
-                    string errorMessage = ProcessChecks(val);
-                    if (errorMessage == null)
-                        return;
+                var optionalAttribute = property.GetCustomAttribute(typeof(OptionalAttribute));
+                
+                if (optionalAttribute == null)
+                    throw new CheckFieldException(property.Name, val, "Empty setting value");
```

Comment:
Is it correct to throw an exception? The call stack was reworked to expect error strings.

Reply:
yes, changed the code to rethrow CheckFieldException